python/metatensor-learn/metatensor/learn/nn/module_map.py

```diff
@@ -0,0 +1,463 @@
+
+        >>> block_1 = TensorBlock(
+        ...     values=torch.tensor(
+        ...         [
+        ...             [1.0, 2.0, 4.0],
+        ...             [3.0, 5.0, 6.0],
+        ...         ]
+        ...     ),
+        ...     samples=Labels(
+        ...         ["structure", "center"],
+        ...         np.array(
+        ...             [
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[0], [1], [2]])),
+        ... )
+        >>> block_2 = TensorBlock(
+        ...     values=torch.tensor(
+        ...         [
+        ...             [5.0, 8.0, 2.0],
+        ...             [1.0, 2.0, 8.0],
+        ...         ]
+        ...     ),
+        ...     samples=Labels(
+        ...         ["structure", "center"],
+        ...         np.array(
+        ...             [
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[3], [4], [5]])),
+        ... )
+        >>> keys = Labels(names=["key"], values=np.array([[0], [1]]))
+        >>> tensor = TensorMap(keys, [block_1, block_2])
+
+        Create modules
+
+        >>> linear = torch.nn.Linear(3, 1, bias=False)
+        >>> with torch.no_grad():
+        ...     _ = linear.weight.copy_(torch.tensor([1.0, 1.0, 1.0]))
+        ...
+        >>> modules = [linear, deepcopy(linear)]
+        >>> # you could also extend the module by some nonlinear activation function
+
+        Create ModuleMap from this ModucDict and apply it
+
+        >>> module_map = ModuleMap(tensor.keys, modules)
+        >>> out = module_map(tensor)
+        >>> out
+        TensorMap with 2 blocks
+        keys: key
+               0
+               1
+        >>> out[0].values
+        tensor([[ 7.],
+                [14.]], grad_fn=<MmBackward0>)
+        >>> out[1].values
+        tensor([[15.],
+                [11.]], grad_fn=<MmBackward0>)
+
+        Lets look at the metadata
+
+        >>> tensor[0]
+        TensorBlock
+            samples (2): ['structure', 'center']
+            components (): []
+            properties (3): ['properties']
+            gradients: None
+        >>> out[0]
+        TensorBlock
+            samples (2): ['structure', 'center']
+            components (): []
+            properties (1): ['_']
+            gradients: None
+
+        It got completely lost because we cannot know in general what the output is.
+        You can add in the initialization of the ModuleMap a TensorMap that contains
+        the intended output Labels.
+    """
+
+    def __init__(
+        self,
+        in_keys: Labels,
+        modules: Sequence[Module],
+        out_properties: Dict[LabelsEntry, Labels] = None
```

Comment:
This might fail with torchscript. I was more thinking of a custom `LabelsEntryDict` class that would store a list of values and Labels. It will be a bit of work, so we could do this in a later PR. For now, maybe `out_properties=Optional[List[Labels]]`, with the same number of elements as keys`?

Reply:
Yes that was a question here https://github.com/lab-cosmo/metatensor/pull/427#discussion_r1457592018 Then I do it this way